acond/src/restful.rs

```diff
@@ -0,0 +1,577 @@
+            let stream = ref_stream.lock().await;
+            let file = File::open(path).await?;
+            stream.send_fd(file.as_raw_fd()).await?;
+            unistd::close(file.as_raw_fd())?;
+            unistd::unlink(path)?;
+        }
+
+        let recv_buf = acond_io::read_async_lock(self.stream.clone()).await?;
+        if recv_buf.is_empty() {
+            return Err(anyhow!(utils::ERR_UNEXPECTED));
+        }
+        Ok(recv_buf)
+    }
+}
+
+fn get_rest_error(buf: &[u8]) -> Result<RestError> {
+    let error: AcondError = bincode::deserialize(buf)?;
+    match error.code {
+        Code::Unknown => Ok(RestError::Unknown(error.message)),
+        Code::InvalidArgument => Ok(RestError::InvalidArgument(error.message)),
+        Code::DeadlineExceeded => Ok(RestError::DeadlineExceeded(error.message)),
+        Code::PermissionDenied => Ok(RestError::PermissionDenied(error.message)),
+    }
+}
+
+pub async fn run_server(
+    stream: StdUnixStream,
+    port: u32,
+) -> Result<(), Box<dyn std::error::Error>> {
+    let service = web::Data::new(ExchangeService::new(UnixStream::from_std(stream)?));
+
+    let rsa_key = Rsa::generate(3072)?;
+    let private_key = PKey::from_rsa(rsa_key)?;
+
+    let mut x509_builder = X509Builder::new()?;
+
+    let start_time = Asn1Time::days_from_now(0)?;
+    let end_time = Asn1Time::days_from_now(365)?;
+    x509_builder.set_not_before(&start_time)?;
+    x509_builder.set_not_after(&end_time)?;
+
+    let mut x509_name_builder = X509NameBuilder::new()?;
+    x509_name_builder.append_entry_by_text("CN", "localhost")?;
+    let subject_name = x509_name_builder.build();
+
+    x509_builder.set_subject_name(&subject_name)?;
+    x509_builder.set_issuer_name(&subject_name)?;
+    x509_builder.set_pubkey(&private_key)?;
+
+    let mut extension = BasicConstraints::new();
+    extension.ca();
+    x509_builder.append_extension(extension.build()?)?;
+
+    x509_builder.sign(&private_key, MessageDigest::sha384())?;
+    let x509 = x509_builder.build();
+
+    HttpServer::new(move || {
+        App::new()
+            .app_data(service.clone())
+            .route("/api/v1/manifest", web::post().to(add_manifest))
+            .route("/api/v1/manifest/finalize", web::post().to(finalize))
+            .route(
+                "/api/v1/manifest/{alg}/{hash1}/{hash2}",
+                web::get().to(get_manifest),
+            )
+            .route("/api/v1/blob/{name}", web::put().to(add_blob))
+            .route("/api/v1/blob/{name}", web::get().to(get_blob_size))
+            .route("/api/v1/container/start", web::post().to(start))
+            .route("/api/v1/container/restart", web::post().to(restart))
+            .route("/api/v1/container/exec", web::post().to(exec))
+            .route("/api/v1/container/kill", web::post().to(kill))
+            .route("/api/v1/container/inspect", web::get().to(inspect))
+            .route("/api/v1/container/report", web::get().to(report))
+    })
+    .workers(10)
```

Comment:
We can take the default number of workers.

Reply:
OK